refine-ecommerce-blog (#1549)
pankod/refine: documentation/blog/2022-02-14-refine-ecommerce-blog.md

---

> title: Refine-Nextjs E-commerce App with Strapi-v4 + Chakra UI

Reviewer: ```suggestion
title: Next.js E-commerce App with Strapi and Chakra UI
```

> description: In this guide, we will develop an E-commerce application using the headless feature of refine.

Reviewer: ```suggestion
description: Learn the power of Refine for e-commerce with this quick & easy example. This step-by-step Refine SPA tutorial will get you started in no time.
```

slug: refine-ecommerce
authors: melih

> tags: [refine, E-commerce, strapi, chakra-ui]

Reviewer: ```suggestion
tags: [refine, e-commerce, strapi, chakra-ui, next.js]
```

//highlight-start
export const getServerSideProps: GetServerSideProps = async (context) => {
    const axiosInstance = axios.create();

    const data = await DataProvider(
        API_URL + "/api",
        axiosInstance,
    ).getList<IProduct>({
        resource: "products",
        metaData: { populate: ["image"] },
    });

    return {
        props: { products: data },
    };
};
//highlight-end
```

### Create Product List with Refine

Let's process the data we fetch above using **refine**'s `useTable` hook. Then let's put our data in our ProductCard component.

```tsx title="pages/index.tsx"
import { GetServerSideProps } from "next";
//highlight-next-line
import { LayoutWrapper, GetListResponse, useTable } from "@pankod/refine-core";
import { DataProvider } from "@pankod/refine-strapi-v4";

import { API_URL } from "src/constants";
import { IProduct } from "interfaces";
import { SimpleGrid } from "@chakra-ui/react";
import { ProductCard } from "src/components";

import axios from "axios";

//highlight-start
type ItemProps = {
    products: GetListResponse<IProduct>;
};
//highlight-end

export const ProductList: React.FC<ItemProps> = ({ products }) => {
    //highlight-start

>     const { tableQueryResult } = useTable<IProduct>({
>         resource: "products",
>         metaData: { populate: ["image"] },
>     });

Reviewer: ```suggestion
    const { tableQueryResult } = useTable<IProduct>({
        resource: "products",
        queryOptions: {
            initialData: products,
         },
        metaData: { populate: ["image"] },
    });
```

        <LayoutWrapper>
            <SimpleGrid columns={[1, 2, 3]} mt={6} spacing={3}>
                {tableQueryResult.data?.data.map((item) => (
                    <ProductCard
                        id={item.id}
                        title={item.title}
                        description={item.description}
                        cardImage={item.image ? API_URL + item.image.url : "./error.png"}
                    />
                ))}
            </SimpleGrid>
        </LayoutWrapper>
        //highlight-end
    );
};

export default ProductList;

export const getServerSideProps: GetServerSideProps = async (context) => {
    const axiosInstance = axios.create();

    const data = await DataProvider(
        API_URL + "/api",
        axiosInstance,
    ).getList<IProduct>({
        resource: "products",
        metaData: { populate: ["image"] },
    });

    return {
        props: { products: data },
    };
};
```

<div class="img-container">
    <div class="window">
        <div class="control red"></div>
        <div class="control orange"></div>
        <div class="control green"></div>
    </div>
    <img src={products} alt="products" />
</div>
<br />

> ## Fetch and List the Stores

Reviewer: ```suggestion
## Adding Store-based Filtering
```

                        description={item.description}
                        cardImage={item.image ? API_URL + item.image.url : "./error.png"}
                    />
                ))}
            </SimpleGrid>
        </LayoutWrapper>
    );
};

export default ProductList;

export const getServerSideProps: GetServerSideProps = async (context) => {
    const axiosInstance = axios.create();

    const data = await DataProvider(
        API_URL + "/api",
        axiosInstance,
    ).getList<IProduct>({
        resource: "products",
        metaData: { populate: ["image"] },
        pagination: { current: 1, pageSize: 9 },
    });

    const { data: storesData } = await DataProvider(
        API_URL + "/api",
        axiosInstance,
    ).getMany({
        resource: "stores",
        ids: ["1", "2", "3"],
    });

    return {
        props: {
            products: data,
            stores: storesData,
        },
    };
};
```

<div class="img-container">
    <div class="window">
        <div class="control red"></div>
        <div class="control orange"></div>
        <div class="control green"></div>
    </div>
    <img src={pagination} alt="pagination" />
</div>
<br />

> ## Set Product Capacity(Pagination)

Reviewer: ```suggestion
## Adding Pagination Capability
```

                        >
                            {"Page: " + (i + 1)}
                        </Button>
                    );
                })}
            </Flex>
            //highlight-end
        </LayoutWrapper>
    );
};

export default ProductList;

export const getServerSideProps: GetServerSideProps = async (context) => {
    const axiosInstance = axios.create();

    const data = await DataProvider(
        API_URL + "/api",
        axiosInstance,
    ).getList<IProduct>({
        resource: "products",
        metaData: { populate: ["image"] },
        pagination: { current: 1, pageSize: 9 },
    });

    const { data: storesData } = await DataProvider(
        API_URL + "/api",
        axiosInstance,
    ).getMany({
        resource: "stores",
        ids: ["1", "2", "3"],
    });

    return {
        props: { products: data, stores: storesData },
    };
};
```

<div class="img-container">
    <div class="window">
        <div class="control red"></div>
        <div class="control orange"></div>
        <div class="control green"></div>
    </div>
    <img src={page_size} alt="page_size" />
</div>
<br />

> ## Add Items to Basket and Payout

Reviewer: ```suggestion
## Adding Basket and Payment capabilities with Snipcart
```

One of the steps that should be in an E-commerce application is the cart and payment transactions. In our example, we will use [Snipcart](https://snipcart.com/) for this process.

[Refer to the Snipcart documentation for detailed information. →](https://docs.snipcart.com/v3/)

> ### Installation

Reviewer: ```suggestion
### Installation Snipcart Widget
```

    const axiosInstance = axios.create();
    const dataProvider = DataProvider(API_URL + "/api", axiosInstance);
    return (
        <>
            //highlight-start
            <Head>
                <link rel="preconnect" href="https://app.snipcart.com" />
                <link
                    rel="stylesheet"
                    href="https://cdn.snipcart.com/themes/v3.0.16/default/snipcart.css"
                />
                <script
                    async
                    src="https://cdn.snipcart.com/themes/v3.0.16/default/snipcart.js"
                />
            </Head>
            //highlight-end
            <Refine
                routerProvider={routerProvider}
                dataProvider={dataProvider}
                resources={[{ name: "products" }]}
                Layout={Layout}
            >
                <ChakraProvider>
                    <Component {...pageProps} />
                </ChakraProvider>
            </Refine>
            //highlight-start
            <div hidden id="snipcart" data-api-key="YOUR_SNIPCART_TEST_KEY" />
            //highlight-end
        </>
    );
}
```

> ### Product Details for Add to Basket Button

Reviewer: ```suggestion
### Adding "Add to Basket" Button on ProductCard component
```

                >
                    //highlight-start
                    <Button
                        className="buy-button snipcart-add-item"
                        bgColor={"green.400"}
                        data-item-id={id}
                        data-item-price="5"
                        data-item-url="/"
                        data-item-name={title}
                        data-item-description={description}
                        data-item-image={cardImage}
                    >
                        Add to Basket
                    </Button>
                    //highlight-end
                </Box>
            </Box>
        </Box>
    );
};
```

<div class="img-container">
    <div class="window">
        <div class="control red"></div>
        <div class="control orange"></div>
        <div class="control green"></div>
    </div>
    <img src={basket} alt="basket" />
</div>
<br />

## Conclusion

One of the biggest features that distinguishes Refine from other frameworks is that it is customizable. Combined with Refine headless, it now provides more customization options. This provides a great deal of convenience in the project you will develop.

> As you can see in this article, we have developed the Client part of our Admin Panel, which we have done before, with **refine**. Refine offers the opportunity to develop B2B and B2C applications without any restrictions and in a fully customizable manner.

Reviewer: ```suggestion
As you can see in this article, we have developed the Client part of our [Admin Panel](https://refine.dev/docs/guides-and-concepts/multi-tenancy/strapi-v4/), which we have done before, with **refine**. Refine offers the opportunity to develop B2B and B2C applications without any restrictions and in a fully customizable manner.

[Refer to the Admin side of the project →](https://refine.dev/docs/guides-and-concepts/multi-tenancy/strapi-v4/)
```

description: Learn the power of Refine for e-commerce with this quick & easy example. This step-by-step Refine SPA tutorial will get you started in no time.
slug: refine-ecommerce
authors: melih
tags: [refine, e-commerce, strapi, chakra-ui, next.js]
image: https://refine.dev/img/refine_social.png
hide_table_of_contents: false
---

import products from '@site/static/img/blog/2022-02-14-refine-ecommerce-blog/products.png';
import pagination from '@site/static/img/blog/2022-02-14-refine-ecommerce-blog/pagination.gif';
import layout from '@site/static/img/blog/2022-02-14-refine-ecommerce-blog/layout.png';
import sample_product from '@site/static/img/blog/2022-02-14-refine-ecommerce-blog/sample_product.png';
import basket from '@site/static/img/blog/2022-02-14-refine-ecommerce-blog/basket.gif';
import page_size from '@site/static/img/blog/2022-02-14-refine-ecommerce-blog/page_size.gif';

It is now used **headless** with the **refine** 3 version. You can use any UI library you want with the **headless** feature.

In this article, we will create the E-commerce client of our [Strapi-Multitenancy](/docs/guides-and-concepts/multi-tenancy/strapi-v4) admin panel that we have done before.

> We will use [Strapi](https://strapi.io/) and [Chakra-UI](https://chakra-ui.com/) together with [**refine-nextjs(SSR)**](/docs/guides-and-concepts/ssr-nextjs/) in our E-commerce client example application.

Reviewer: ```suggestion
We will use [Strapi](https://strapi.io/) and [Chakra-UI](https://chakra-ui.com/) together with [**Next.js**](/docs/guides-and-concepts/ssr-nextjs/) in our E-commerce client example application.
```

<!--truncate-->

## Refine Project Setup

Let's start by creating our **refine** project. You can use the [superplate](https://github.com/pankod/superplate) to create a refine project.

```bash
npx superplate-cli -p refine-nextjs refine-ecommerce-example
```

```bash
✔ What will be the name of your app · refine-ecommerce-example
✔ Package manager: · npm
✔ Do you want to using UI Framework? > No(headless)
✔ Data Provider: Strapi
✔ i18n - Internationalization: · no
```

superplate will quickly create our **refine** project according to the features we choose. Let's continue by install the [**refine** Strapi-v4 Data Provider](/docs/guides-and-concepts/data-provider/strapi-v4/) and Chakra-UI packages that we will use later.

## Installation

```bash
cd refine-ecommerce-example

> npm i @pankod/refine-strapi-v4
> npm i @chakra-ui/react @emotion/react@^11 @emotion/styled@^11 framer-motion@^6

Reviewer: ```suggestion
npm i @pankod/refine-strapi-v4 @chakra-ui/react @emotion/react@^11 @emotion/styled@^11 framer-motion@^6
```

```

Our **refine** project and installations are now ready! Let's start using it.

## Usage

> ### Define Strapi-v4 Provider

Reviewer: ```suggestion
### Configure Refine for Strapi-v4
```

```tsx title="pages/_app.tsx"
import React from "react";
import { AppProps } from "next/app";
import Head from "next/head";
import { Refine } from "@pankod/refine-core";
import routerProvider from "@pankod/refine-nextjs-router";
//highlight-next-line
import { DataProvider } from "@pankod/refine-strapi-v4";

> import axios from "axios";

Reviewer: ```suggestion
```

> const API_URL = "YOUR_STRAPI_API_URL";

Reviewer: ```suggestion
const API_URL = "https://api.strapi-multi-tenant.refine.dev/api";
```

function MyApp({ Component, pageProps }: AppProps): JSX.Element {

>     const axiosInstance = axios.create();

Reviewer: ```suggestion
```

    //highlight-next-line

>     const dataProvider = DataProvider(API_URL + "/api", axiosInstance);

Reviewer: ```suggestion
    const dataProvider = DataProvider(API_URL);
```

    return (
        <Refine
            routerProvider={routerProvider}
            //highlight-next-line
            dataProvider={dataProvider}
        >
            <Component {...pageProps} />
        </Refine>
    );
}
```

> ### Wrap the Component with ChakraProvider

Reviewer: ```suggestion
### Chakra-UI Provider Setup
```

```tsx title="pages/_app.tsx"
import React from "react";
import { AppProps } from "next/app";
import Head from "next/head";
import { Refine } from "@pankod/refine-core";
import routerProvider from "@pankod/refine-nextjs-router";
import { DataProvider } from "@pankod/refine-strapi-v4";

//highlight-next-line
import { ChakraProvider } from "@chakra-ui/react";

> const API_URL = "YOUR_STRAPI_API_URL";

Reviewer: ```suggestion
const API_URL = "https://api.strapi-multi-tenant.refine.dev/api";
```

> import axios from "axios";

Reviewer: ```suggestion
```

function MyApp({ Component, pageProps }: AppProps): JSX.Element {

>     const axiosInstance = axios.create();
>     const dataProvider = DataProvider(API_URL + "/api", axiosInstance);

Reviewer: ```suggestion
    const dataProvider = DataProvider(API_URL + "/api");
```

    return (
        <Refine routerProvider={routerProvider} dataProvider={dataProvider}>
            //highlight-start
            <ChakraProvider>
                <Component {...pageProps} />
            </ChakraProvider>
            //highlight-end
        </Refine>
    );
}
```

## Create Strapi Collections

We created three collections on Strapi as `store`, `product` and `order` and added a relation between them. For detailed information on how to create a collection, you can check [here](https://docs.strapi.io/developer-docs/latest/getting-started/quick-start.html).

We created our collections in the previous Strapi Multitenancy guide. Now we will use the same collections.

[Refer to the Project Collections for detailed information. →](https://refine.dev/docs/guides-and-concepts/multi-tenancy/strapi-v4/#create-collections)

> ## Create Layout

Reviewer: ```suggestion
## Create Refine Layout
```
